lib/cookie.c

```diff
@@ -217,6 +240,49 @@ static bool pathmatch(const char *cookie_path, const char *request_uri)
   return ret;
 }
 
+/*
+ * Return the top-level domain, for optimal hashing.
+ */
+static const char *get_top_domain(const char * const domain, size_t *outlen)
+{
+  size_t len;
+  const char *first, *last;
+
+  if(!domain)
+    return NULL;
+
+  len = strlen(domain);
+  first = memchr(domain, '.', len);
+  last = memrchr(domain, '.', len);
+
+  if(outlen)
+    *outlen = len;
+
+  if(first == last)
+    return domain;
+
+  first = memrchr(domain, '.', (size_t)(last - domain - 1));
+  if(outlen)
+    *outlen = len - (size_t)(first - domain) - 1;
+
+  return first + 1;
+}
+
+/*
+ * Hash this domain.
+ */
+static size_t cookiehash(const char * const domain)
+{
+  const char *top;
+  size_t len;
+
+  if(!domain || isip(domain))
+    return 0;
+
+  top = get_top_domain(domain, &len);
+  return Curl_hash_str((void *) top, len, COOKIE_HASH_SIZE);
```

Comment:
Should compute hash on lowercase domain to avoid missing a match in later strcasecompare().